### ISSUE TITLE
Expression of type "List[None]" cannot be assigned to declared type "List[int | None]"

### ISSUE THREAD

**Issue by ozars** [microsoft/pyright#1691]
**Describe the bug**
Multiplicated list of `None`s cannot be assigned to list of a type unioned with `None` even with explicit type signature.

**To Reproduce**
```python
from typing import List, Optional, Union
x: Union[int, None] = None                # ok
lst0: List[Optional[int]] = [None, None]  # ok
lst1: List[Optional[int]] = [None] * 2    # fail
```

```
No configuration file found.
stubPath /home/omer/test/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/omer/test/test.py
  /home/omer/test/test.py:3:28 - error: Expression of type "List[None]" cannot be assigned to declared type "List[int | None]"
    TypeVar "_T@list" is invariant
      Type "None" cannot be assigned to type "int | None"
        Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.561sec
```

**Expected behavior**
No type error.

**VS Code extension or command-line**
cli pyright 1.1.125

**Comment by erictraut** [comment]
This is a duplicate of https://github.com/microsoft/pyright/issues/1073.

The last time I looked into this, I didn't see any way to handle it without embedding some special-case knowledge of the semantics of the list class. We generally try very hard to avoid embedding knowledge of specific class semantics in the type checker because this approach isn't scalable. This might be a use case that's common enough for us to bend this rule. Mypy appears to handle this case, which bolsters the argument in favor of special-case handling in pyright.

I'm going to give this some more thought.

**Comment by ozars** [comment]
Thanks. Based on https://github.com/microsoft/pyright/issues/1073#issuecomment-703726470, seems like it's more complex than it looks on the surface.

**Comment by erictraut** [comment]
After some internal discussion with my colleagues, we decided to make an exception for this particular case. I've added special-case logic that specifically handles the situation where 1) the assigned variable has an annotated type that is a built-in list (but not a union that contains a list or a TypeVar that is bound to a list), 2) the assigned expression involves a multiply operator, and 3) the LHS operand is a list expression (using square brackets). This should cover the most common case, but there are many edge cases where it will not work.

Here's the change, if you're interested in the details: https://github.com/microsoft/pyright/commit/a26720b766c6e875707fd0d37c65fd77778b2ac9#diff-e8178cbc8c6f5764a3b368c69c7aad9132af94c62d25559c433c80d58e9cf51d

This change will appear in the next release of pylance.

**Comment by ozars** [comment]
Handling `[None] * x` is likely good enough for covering widespread pattern for preallocating some capacity in a list. Thank you.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.126, which I just published. It will also be included in the next release of pylance.
